homeassistant/components/mqtt/update.py

```diff
@@ -0,0 +1,206 @@
+    )
+    await async_setup_entry_helper(hass, update.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_entity(
+    hass: HomeAssistant,
+    async_add_entities: AddEntitiesCallback,
+    config: ConfigType,
+    config_entry: ConfigEntry,
+    discovery_data: dict[str, Any] | None = None,
+) -> None:
+    """Set up the MQTT update."""
+    async_add_entities([MqttUpdate(hass, config, config_entry, discovery_data)])
+
+
+class MqttUpdate(MqttEntity, UpdateEntity, RestoreEntity):
+    """representation of an MQTT update."""
+
+    _entity_id_format = update.ENTITY_ID_FORMAT
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config: ConfigType,
+        config_entry: ConfigEntry,
+        discovery_data: dict[str, Any] | None = None,
+    ) -> None:
+        """Initialize the MQTT update."""
+        self._config = config
+        self._sub_state = None
+
+        self._attr_device_class = self._config.get(CONF_DEVICE_CLASS)
+
+        UpdateEntity.__init__(self)
+        MqttEntity.__init__(self, hass, config, config_entry, discovery_data)
+
+    @staticmethod
+    def config_schema() -> vol.Schema:
+        """Return the config schema."""
+        return DISCOVERY_SCHEMA
+
+    def _setup_from_config(self, config: ConfigType | dict[str, Any]) -> None:
+        """(Re)Setup the entity."""
+        self._templates = {
+            CONF_VALUE_TEMPLATE: MqttValueTemplate(
+                config.get(CONF_VALUE_TEMPLATE),
+                entity=self,
+            ).async_render_with_possible_json_value,
+            CONF_LATEST_VERSION_TEMPLATE: MqttValueTemplate(
+                config.get(CONF_LATEST_VERSION_TEMPLATE),
+                entity=self,
+            ).async_render_with_possible_json_value,
+        }
+
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics: dict[str, Any] = {}
+
+        def add_subscription(
+            topics: dict[str, Any], topic: str, msg_callback: Callable
+        ) -> None:
+            if self._config.get(topic) is not None:
+                topics[topic] = {
+                    "topic": self._config[topic],
+                    "msg_callback": msg_callback,
+                    "qos": self._config[CONF_QOS],
+                    "encoding": self._config[CONF_ENCODING] or None,
+                }
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_installed_version_received(msg: ReceiveMessage) -> None:
+            """Handle receiving installed version via MQTT."""
+            installed_version = self._templates[CONF_VALUE_TEMPLATE](msg.payload)
+            if isinstance(installed_version, str):
+                self._attr_installed_version = installed_version
+            else:
+                self._attr_installed_version = installed_version.decode(
+                    DEFAULT_ENCODING
+                )
```

Comment:
Guys, I'm not sure this part. Maybe there is a smarter way to reconcile `ReceivePayloadType` with `str`?

Reply:
ReceivePayloadType is the correct type, `CONF_ENCODING` will be used to decode the message. if  `CONF_ENCODING` is `None` the raw bytes will be passed to the function. In any case the result of the template should be a string to be processed, If the result is an empty string, the result payload should be ignored.